homeassistant/components/minecraft_server/__init__.py

```diff
@@ -121,6 +121,24 @@ async def async_migrate_entry(hass: HomeAssistant, config_entry: ConfigEntry) ->
 
         _LOGGER.debug("Migration to version 3 successful")
 
+    # 3 --> 4: Add server type.
+    if config_entry.version == 3:
+        _LOGGER.debug("Migrating from version 3")
+
+        # Existing server(s) with version 2 can only be of type Java Edition.
+        _LOGGER.debug(
+            "Migrating config entry, adding server type '%s'",
+            MinecraftServerType.JAVA_EDITION,
+        )
+
+        config_data = config_entry.data
+        new_data = config_data.copy()
+        new_data[CONF_TYPE] = MinecraftServerType.JAVA_EDITION
+        config_entry.version = 4
+        hass.config_entries.async_update_entry(config_entry, data=new_data)
+
+        _LOGGER.debug("Migration to version 4 successful")
+
```

Comment:
I see you love migrations, but they still aren't backwards compatible. Since we're now only adding a server type, can't we just make the code use Java as fallback?

Reply:
Thanks so much for your VERY fast review :smile: 

Why aren't they backwards compatible? I think I didn't get this point.

The problem would be long update times, because the API currently tries to fetch data 3 times with 3s timeout each. That's why I thought it's the best to store the server type to get fetch the data faster.

Reply:
If you check the code I linked, you can do something like,

```python
if CONF_TYPE not in entry.data:
   -- the place where you check what kind of server it is --
  hass.config_entries.async_update_entry(entry, data=your_new_data)
```

Reply:
Done. Reverted version 4 and moved migration from async_migrate_entry to async_setup_entry.